### ISSUE TITLE
Ordered list only works up to 2

### ISSUE THREAD

**Issue by hiren1610** [pytesNET/tail.writer#4]
First of all, thank you for working on this, this is the only true textarea editor I have found after searching everywhere I can.

I am trying to use tail.writer 0.3.2 - Alpha.

Below are the points I would like to see in the future release or if you can give me some idea to edit the current code and make it work for me.

1. In my project, I would like to have ordered list where user should be able to click the Ordered List button once and then should keep writing the list until he decides to end it by either pressing "Enter" twice to create two empty lines or by pressing on the Ordered List button again. It currently doesn't do that.

2. When I click Ordered List button, I cans see 1. and I can write the line. When I hit enter it creates new line with 2. and I can write one more line. When I hit enter after that, it will end the Ordered list and I have to click on the Ordered List button again to get 3. and I can continue writing the line.

3. It should be able to detect if the last Ordered List has been ended with extra space or if any other new lines are after that then it should start the number from 1. again. instead of continuing the numbers from the previous list. This gives me creating multiple list in the editor.

Please get back to me if you need any clarification or any screenshots to understand what I mean.

Thanks.

**Comment by SamBrishes** [comment]
Hello,

thanks, I'm really glad that you like my script!

To your **first** issue: The current version (should) already offer the possibility to execute a list "walker" by pressing "Enter" twice. BUT, this doesn't work on Ordered lists due to a bug. Your idea to "exit" a walker by pressing the button again is really good .

Your **second** issue is a bug. The walker mechanism currently checks only if the previous line "starts" with the default markup syntax (which is 1. on ordered lists) without applying the markup filter (which turns additional "1."s into ascending numbers).

To your **third** Issue: Yeah, that should be applied too, of course.

Thanks for your ideas, I will implement and fix your issues in the next major release. The current progress is available on [Issue #3](#3).

## Hotfix
To fix the (second) bug with the ordered lists you could add the following snippet to the `.walkable()` method ([on line 856](https://github.com/pytesNET/tail.writer/blob/master/js/tail.writer.js#L856))

```javascript
    walkable: function(string){
        if(!this.walk || string == undefined){
            this.walk = false;
            return false;
        }
        var action = _action(this.walk[0], this.walk[1]),
            syntax = this.indenter(action.markup.replace("$1", ""));


        // @start Ordered Markup Fix
        if(this.walk[0] == "list" && this.walk[1][0] == "ordered"){
            if(/^[0-9]+\.\s+\w+/.test(string)){
                return true;
            }
            if(/^[0-9]+\.\s+/.test(string)){
                this.writeLine("", "replace");
            }
            this.walk = false;
            return false;
        }
        // @end Ordered Markup Fix


        // Check Lines
        if(string.indexOf(syntax) < 0){
            this.walk = false;
            return false;
        }
        if(string == syntax){
            this.writeLine("", "replace");
            this.walk = false;
            return false;
        }
        return true;
    },
```

Sincerely,
Sam.

**Comment by hiren1610** [comment]
Thanks for the quick response and hotfix.

1. Yes that idea can be applied to any sort of list in the project, clicking the button again should exit the list too.

2. hotfix definitely fixed the issue. 💯 %

3. Yup, if that can be fixed in future release then I am waiting for it... eagerly. :)

Regards,
Hiren.

**Comment by SamBrishes** [comment]
Hello,

It would be cool to see your modifications, but no Git/GitHub doesn't has any PM system. You can either post / upload your files directly on this issue, you can also fork this repo to create your own one (and made your modifications there) or you can send me an eMail ([info@pytes.net](mailto:info@pytes.net?subject=tail.writer)), if you want to keep this private.

Unfortunately, I forgot to write an API documentation for ver.0.3.x, and the coming major release will also contain a completely new Interface. But some infos:

```javascript
tail.writer.actions["myaction"] = {
    syntax: "SYNTAX",
    callback: "inline",
    filter: function(before, content, after, selection){
    
    },
    walker: false
}
```

### syntax
This property can contain a STRING like `"[$1](url)"`, where `$1` gets replaced by an selection. You can also disable the syntax by using `false` or you pass an own function. The function gets passed the additional Markup Parameters as arguments: For Example: 
`myaction:param1,param2` = `syntax: function(param1, param2){}`.

### callback
Use either "inline" or "block" (for an respective inline or block action) or pass a function, to 
create your very own interaction. This function receives 4 arguments: "markup" (the syntax), 
"action" (the markup action object), "type" (the markup action name) and "args" (which contains 
all used markup settings). Use `this` to access the main prototype class.

### filter
This callback function allows you to filter the respective content, which has been made after using 
an action. It receives 4 arguments: "before" (contains the content before), "content" 
(contains the current content, already with the Syntax included), "after" (contains the content after) 
and "selection" (contains the current selection object, which can be "manipulated" if you add or remove
some main content). The filter should ALWAYS return the "final" content.

### walker
This property allows just a boolean state. Use `true` if your action is walkable, `false` if not.

------

But as mentioned above, this API will get completely replaces by an easier one.

Sincerely,
Sam.

**Comment by hiren1610** [comment]
Thanks for the reply.

Looks like its still a bit hard for me to develop my own button and add it to the toolbar. I want to call function written in my .js file.
For the moment I will skip this and provide customer buttons outside the editor to make it work. I will wait for the next release for easier API.

**Comment by SamBrishes** [comment]
Hellow,

nope, the CTRL + X key-shortcut does currently only "remove" a line (or selection), it doesn't cut it. I had some issues due to the JavaScript Clipboard functionality.

This will be fixed / added in 0.4.0 too.

**Comment by hiren1610** [comment]
Thanks for the confirmation. I have removed the Cut code from the source for now and provided external button for my project. As I mentioned above, I will wait for the next release for API as well so I will work on the rest of the functionality later on. I am closing this issue for now. :)
